### ISSUE TITLE
OV7251 fails probe, causing cameras to not show up

### ISSUE THREAD

**Issue by qzed** [linux-surface/linux-surface#893]
On my Surface Book 2 using kernel 5.19 (https://github.com/linux-surface/kernel/tree/v5.19-surface-devel), the OV7251 sensor driver fails to probe with
```
[    2.738196] ipu3-cio2 0000:00:14.3: Found supported sensor INT33BE:00
[    2.738341] ipu3-cio2 0000:00:14.3: Found supported sensor INT347A:00
[    2.738406] ipu3-cio2 0000:00:14.3: Found supported sensor INT347E:00
[    2.738430] ipu3-cio2 0000:00:14.3: Connected 3 cameras
...
[    2.748018] ov7251 i2c-INT347E:00: cannot get enable gpio
[    2.748179] ov7251: probe of i2c-INT347E:00 failed with error -2

```

And since ipu3-cio (I think) is waiting on all sensors to show up, no camera is working.

CC @djrscally, I might be missing some things you sent in upstream, or have some outdated patches in the tree. Would be great if you could have a look at this.

**Comment by djrscally** [comment]
Morning - sorry, seen this now. Let me take a loot

**Comment by qzed** [comment]
Let me know if you need me to test anything.

**Comment by skgsergio** [comment]
Surface Pro 5 with Arch Linux here. Let me know if I can help debugging :)

Just tested the camera using `qcam`, and can observe a crash in kernel logs.

- `kernel`: `5.19.2-arch1-1-surface`
- `libcamera`: `3fa79cc6`

<details>
<summary>dmesg</summary>

```
[   38.528977] BUG: kernel NULL pointer dereference, address: 0000000000000028
[   38.528992] #PF: supervisor read access in kernel mode
[   38.528999] #PF: error_code(0x0000) - not-present page
[   38.529005] PGD 0 P4D 0 
[   38.529015] Oops: 0000 [#1] PREEMPT SMP PTI
[   38.529025] CPU: 2 PID: 2458 Comm: qcam Tainted: G         C        5.19.2-arch1-1-surface #1 86ff9efc60c971e718b3e9198747b3c4c0a4086f
[   38.529037] Hardware name: Microsoft Corporation Surface Pro/Surface Pro, BIOS 237.3982.768 12.14.2021
[   38.529042] RIP: 0010:imgu_subdev_set_selection+0xed/0x110 [ipu3_imgu]
[   38.529073] Code: e8 c8 e9 13 d0 5e 5f e9 50 ff ff ff 48 8b 43 10 48 8b 53 18 49 89 04 24 31 c0 49 89 54 24 08 eb 87 66 41 83 7c 24 38 00 74 1b <49> 8b 45 28 49 81 c4 68 02 00 00 48 83 c0 30 e9 53 ff ff ff 0f 0b
[   38.529081] RSP: 0018:ffffb5014f8dbc38 EFLAGS: 00010206
[   38.529090] RAX: 0000000000000000 RBX: ffffb5014f8dbd88 RCX: 0000000000000000
[   38.529096] RDX: ffffb5014f8dbd88 RSI: 0000000000000000 RDI: ffff96930d2e33a0
[   38.529102] RBP: ffffb5014f8dbc58 R08: ffffffffc05dd480 R09: 0000000000000000
[   38.529108] R10: 0000000000000000 R11: 0000000000000001 R12: ffff96930d2e33a0
[   38.529113] R13: 0000000000000000 R14: ffffb5014f8dbd88 R15: ffff96930d2e33a0
[   38.529119] FS:  00007ff1d7fff6c0(0000) GS:ffff969467500000(0000) knlGS:0000000000000000
[   38.529127] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   38.529134] CR2: 0000000000000028 CR3: 000000010adb0001 CR4: 00000000003706e0
[   38.529141] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[   38.529146] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[   38.529152] Call Trace:
[   38.529159]  <TASK>
[   38.529170]  call_set_selection_state+0x73/0x100 [videodev e09c668d3eb6a61e71cbe1149fae9e31a4af6147]`
[   38.529223]  subdev_do_ioctl+0x963/0xf30 [videodev e09c668d3eb6a61e71cbe1149fae9e31a4af6147]
[   38.529273]  subdev_do_ioctl_lock+0x130/0x190 [videodev e09c668d3eb6a61e71cbe1149fae9e31a4af6147]
[   38.529317]  video_usercopy+0x2c4/0x760 [videodev e09c668d3eb6a61e71cbe1149fae9e31a4af6147]
[   38.529357]  ? subdev_do_ioctl+0xf30/0xf30 [videodev e09c668d3eb6a61e71cbe1149fae9e31a4af6147]
[   38.529408]  v4l2_ioctl+0x4a/0x60 [videodev e09c668d3eb6a61e71cbe1149fae9e31a4af6147]
[   38.529445]  __x64_sys_ioctl+0x91/0xd0
[   38.529461]  do_syscall_64+0x5c/0x90
[   38.529482]  ? syscall_exit_to_user_mode+0x1b/0x40
[   38.529492]  ? do_syscall_64+0x6b/0x90
[   38.529504]  entry_SYSCALL_64_after_hwframe+0x63/0xcd
[   38.529516] RIP: 0033:0x7ff1f8b159ef
[   38.529525] Code: 00 48 89 44 24 18 31 c0 48 8d 44 24 60 c7 04 24 10 00 00 00 48 89 44 24 08 48 8d 44 24 20 48 89 44 24 10 b8 10 00 00 00 0f 05 <89> c2 3d 00 f0 ff ff 77 18 48 8b 44 24 18 64 48 2b 04 25 28 00 00
[   38.529532] RSP: 002b:00007ff1d7ffd6c0 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
[   38.529541] RAX: ffffffffffffffda RBX: 0000555b9b7365d0 RCX: 00007ff1f8b159ef
[   38.529547] RDX: 00007ff1d7ffd780 RSI: 00000000c040563e RDI: 0000000000000019
[   38.529552] RBP: 00007ff1d7ffd740 R08: 00007ff1d003fd90 R09: 00007ff1d7ffd870
[   38.529557] R10: 0000000000000056 R11: 0000000000000246 R12: 0000000000000001
[   38.529562] R13: 0000000000000000 R14: 00007ffd89706357 R15: 00007ff1d77ff000
[   38.529577]  </TASK>
[   38.529581] Modules linked in: exfat overlay snd_seq_dummy snd_hrtimer rfcomm snd_seq snd_seq_device cmac algif_hash algif_skcipher af_alg bnep btusb btrtl btbcm btintel btmtk bluetooth ecdh_generic xt_MASQUERADE xt_conntrack xt_CHECKSUM ipt_REJECT nf_reject_ipv4 snd_soc_avs hid_sensor_als hid_sensor_gyro_3d snd_soc_skl hid_sensor_rotation surface_platform_profile hid_sensor_accel_3d hid_sensor_trigger industrialio_triggered_buffer kfifo_buf hid_sensor_iio_common industrialio uinput platform_profile xt_tcpudp snd_soc_hdac_hda joydev mousedev snd_hda_ext_core snd_soc_sst_ipc snd_soc_sst_dsp intel_tcc_cooling snd_soc_acpi_intel_match x86_pkg_temp_thermal snd_soc_acpi intel_powerclamp hid_sensor_hub snd_hda_codec_hdmi coretemp snd_soc_core ip6table_mangle kvm_intel snd_compress 8250_dw ac97_bus mei_pxp ip6table_nat snd_hda_codec_realtek ipts snd_pcm_dmaengine snd_hda_codec_generic ledtrig_audio mei_hdcp ip6table_filter ip6_tables kvm irqbypass snd_hda_intel snd_intel_dspcfg
[   38.529742]  crct10dif_pclmul crc32_pclmul ghash_clmulni_intel snd_intel_sdw_acpi iptable_mangle hid_multitouch mac_hid aesni_intel snd_hda_codec iptable_nat nf_nat snd_hda_core snd_hwdep nf_conntrack crypto_simd intel_ishtp_hid cryptd nf_defrag_ipv6 mwifiex_pcie snd_pcm nf_defrag_ipv4 snd_timer libcrc32c gpio_keys intel_rapl_msr mwifiex iptable_filter rapl snd surface_gpe intel_cstate intel_uncore pcspkr uas mei_me soundcore cfg80211 bridge mei intel_lpss_pci intel_xhci_usb_role_switch stp intel_lpss intel_skl_int3472_tps68470 llc usb_storage usbhid ipu3_cio2 rfkill tps68470_regulator idma64 ov5693 clk_tps68470 roles intel_pch_thermal ov7251 ov8865 tpm_crb v4l2_fwnode intel_skl_int3472_discrete v4l2_async tpm_tis i915 tpm_tis_core ipu3_imgu(C) tpm drm_buddy processor_thermal_device_pci_legacy videobuf2_dma_sg surface_aggregator_registry vfat surface_acpi_notify processor_thermal_device rng_core videobuf2_memops video fat processor_thermal_rfim surfacepro3_button surface_aggregator
[   38.529915]  soc_button_array videobuf2_v4l2 processor_thermal_mbox ttm int3400_thermal processor_thermal_rapl videobuf2_common dptf_power drm_display_helper int3403_thermal acpi_thermal_rel acpi_tad intel_rapl_common cec intel_ish_ipc acpi_pad wmi int340x_thermal_zone videodev intel_soc_dts_iosf intel_ishtp mc intel_gtt lzo_rle usbip_host usbip_core dm_multipath dm_mod sg crypto_user fuse zram ip_tables x_tables ext4 crc32c_generic crc16 mbcache jbd2 nvme xhci_pci nvme_core crc32c_intel xhci_pci_renesas
[   38.530010] Unloaded tainted modules: acpi_cpufreq():1 acpi_cpufreq():1 acpi_cpufreq():1 acpi_cpufreq():1 pcc_cpufreq():1 pcc_cpufreq():1 pcc_cpufreq():1 acpi_cpufreq():1 acpi_cpufreq():1 fjes():1 acpi_cpufreq():1 pcc_cpufreq():1 pcc_cpufreq():1 fjes():1 acpi_cpufreq():1 fjes():1 pcc_cpufreq():1 acpi_cpufreq():1 pcc_cpufreq():1 fjes():1 acpi_cpufreq():1 pcc_cpufreq():1 acpi_cpufreq():1 fjes():1 pcc_cpufreq():1 acpi_cpufreq():1 fjes():1 pcc_cpufreq():1 acpi_cpufreq():1 pcc_cpufreq():1 fjes():1 acpi_cpufreq():1 pcc_cpufreq():1 acpi_cpufreq():1 pcc_cpufreq():1 acpi_cpufreq():1 pcc_cpufreq():1 acpi_cpufreq():1 pcc_cpufreq():1 acpi_cpufreq():1 pcc_cpufreq():1 acpi_cpufreq():1 acpi_cpufreq():1
[   38.530136] CR2: 0000000000000028
[   38.530142] ---[ end trace 0000000000000000 ]---
[   38.530147] RIP: 0010:imgu_subdev_set_selection+0xed/0x110 [ipu3_imgu]
[   38.530173] Code: e8 c8 e9 13 d0 5e 5f e9 50 ff ff ff 48 8b 43 10 48 8b 53 18 49 89 04 24 31 c0 49 89 54 24 08 eb 87 66 41 83 7c 24 38 00 74 1b <49> 8b 45 28 49 81 c4 68 02 00 00 48 83 c0 30 e9 53 ff ff ff 0f 0b
[   38.530180] RSP: 0018:ffffb5014f8dbc38 EFLAGS: 00010206
[   38.530187] RAX: 0000000000000000 RBX: ffffb5014f8dbd88 RCX: 0000000000000000
[   38.530192] RDX: ffffb5014f8dbd88 RSI: 0000000000000000 RDI: ffff96930d2e33a0
[   38.530197] RBP: ffffb5014f8dbc58 R08: ffffffffc05dd480 R09: 0000000000000000
[   38.530202] R10: 0000000000000000 R11: 0000000000000001 R12: ffff96930d2e33a0
[   38.530207] R13: 0000000000000000 R14: ffffb5014f8dbd88 R15: ffff96930d2e33a0
[   38.530213] FS:  00007ff1d7fff6c0(0000) GS:ffff969467500000(0000) knlGS:0000000000000000
[   38.530220] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   38.530225] CR2: 0000000000000028 CR3: 000000010adb0001 CR4: 00000000003706e0
[   38.530232] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[   38.530236] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
```
</details>

<details>
<summary>lsmod | grep -E "(ov|ipu3)"</summary>

```
overlay               180224  0
ipu3_cio2              57344  0
ov5693                 28672  1
ov7251                 36864  0
ov8865                 40960  1
v4l2_fwnode            32768  4 ipu3_cio2,ov7251,ov5693,ov8865
v4l2_async             32768  5 v4l2_fwnode,ipu3_cio2,ov7251,ov5693,ov8865
ipu3_imgu             241664  2
videobuf2_dma_sg       20480  2 ipu3_cio2,ipu3_imgu
videobuf2_v4l2         40960  2 ipu3_cio2,ipu3_imgu
videobuf2_common       86016  5 ipu3_cio2,videobuf2_v4l2,videobuf2_dma_sg,ipu3_imgu,videobuf2_memops
videodev              307200  10 v4l2_async,ipu3_cio2,ov7251,videobuf2_v4l2,ov5693,videobuf2_common,ov8865,ipu3_imgu
mc                     69632  9 v4l2_async,videodev,ipu3_cio2,ov7251,videobuf2_v4l2,ov5693,videobuf2_common,ov8865,ipu3_imgu
```
</details>

**Comment by djrscally** [comment]
OK I know what's missing - there's a patch series that almost made it upstream but someone asked for a last minute one line change, which I keep outright forgetting to do :). I'll port it to 5.19 and raise a PR here, and then also actually make the effort to get it upstream.

@skgsergio yours is something else - can we have a separate issue for it please?

**Comment by skgsergio** [comment]
For sure, will open a new one. I thought it could be related.

**Comment by qzed** [comment]
I can confirm that in 5.19.4, the cameras do show up even though ov7251 still fails to probe. I'm not entirely sure what changed though, the only commit related to those parts that I could find between 5.19 and 5.19.2 is https://github.com/linux-surface/kernel/commit/3614c1981fe2.

**Comment by 2ndBillGates** [comment]
On my Surface Book 2 15"  running Fedora 36 `6.0.8-3.surface.fc36.x86_64`
The back Camera is not working. "OV8865"

> [0:17:29.279416823] [24682] ERROR V4L2 v4l2_device.cpp:91 'dw9719 3-000c': Failed to open V4L2 device: No such file or directory
> 
> [0:17:29.279430423] [24682] ERROR CameraSensor camera_sensor.cpp:469 'ov8865 3-0010': CameraLens initialisation failed
> 
> [0:17:29.281057000] [24682] ERROR IPAProxy ipa_proxy.cpp:149 Configuration file 'ov5693.yaml' not found for IPA module 'ipu3'
